parlai/core/torch_agent.py

```diff
@@ -789,10 +814,25 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None):
             # turn on amsgrad for adam
             # amsgrad paper: https://openreview.net/forum?id=ryQu7f-RZ
             kwargs['amsgrad'] = True
+            if self.fp16 and self.fp16_impl == 'mem_efficient':
+                # grab this implementation instead
+                opt['optimizer'] = 'mem_eff_adam'
         elif opt['optimizer'] == 'qhadam':
             # set nus for qhadam
             kwargs['nus'] = opt.get('nus', (0.7, 1.0))
-        if opt['optimizer'] in ['adam', 'sparseadam', 'fused_adam', 'adamax', 'qhadam']:
+        elif opt['optimizer'] == 'adafactor':
+            # adafactor params
+            kwargs['beta1'] = opt.get('betas', (0.9, 0.999))[0]
+            kwargs['eps'] = opt['adafactor_eps']
+            kwargs['warmup_init'] = opt.get('warmup_updates', -1) > 0
+
+        if opt['optimizer'] in [
+            'adam',
+            'sparseadam',
+            'fused_adam',
+            'adamax',
+            'qhadam',
```

Comment:
we don't need to set `betas` with the `mem_eff_adam`?

Reply:
no, instead we take the first beta only